### PR TITLE
KTOR-7644 Make re-auth status codes configurable

### PR DESCRIPTION
**Pull request by wkornewald**
Some services use 403 instead of 401. Changing them might be impossible. With this change Ktor can flexibly work with any broken service.

[KTOR-7644](https://youtrack.jetbrains.com/issue/KTOR-7644) Auth: Make re-auth/refresh status codes configurable

### CODE REVIEW

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt
Comment (osipxd) [inline ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt:141]: Could you also change log messages to reflect the changes and replace `.trace(...)` call with `.trace { ... }` in this file?

**Review by osipxd — COMMENTED** [review]
Thank you for the pull request!
Let's discuss an alternative approach before we merge this.

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt
Comment (wkornewald) [reply ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt]: This sounds much better than the list. I've pushed that change.

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt
Comment (wkornewald) [reply ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt]: Sure. Done in the isUnauthorized commit.

**Review by osipxd — APPROVED** [review]
LGTM. Thank you!
@bjhham, @marychatte, could you also take a look?

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt
Comment (wkornewald) [inline ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt:45]: Something else: this revers to variables that are actually defined on AuthConfig. Shouldn't the docstring be defined there instead or at least refer to the full path `[AuthConfig.providers]`? I just copied the existing code, but maybe this should be fixed, too.

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt
Comment (marychatte) [reply ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt]: @osipxd 
I like the idea with lambda!

> returning 200 OK with actual status code in a body

Do you mean it could be used like this?
 `isUnauthorized = { it.bodyAsText() == "401" }`
 If yes, then `isUnauthorized` function should be suspend

**Comment by marychatte** [comment]
Thanks for the PR! Can we also please add tests for it?

**Comment by wkornewald** [comment]
> Thanks for the PR! Can we also please add tests for it?

Sure, added one and switched to a suspend lambda.

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt
Comment (osipxd) [inline ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt:40]: Let's refine the API a bit to make it more consistent to the existing ones.

```suggestion
    internal var isUnauthorized: suspend (HttpResponse) -> Boolean = { it.status == HttpStatusCode.Unauthorized }

    /** ... */
    public fun isUnauthorized(block: suspend (HttpResponse) -> Boolean) {
        isUnauthorized = block
    }
```

Though, I'm not sure about naming. Probably, `isUnauthorizedResponse` would be better. @marychatte, what do you think?

File: buildSrc/src/main/kotlin/test/server/tests/Auth.kt
Comment (osipxd) [inline buildSrc/src/main/kotlin/test/server/tests/Auth.kt:132]: Nit:
```suggestion
                        val status = call.request.queryParameters["status"]?.toIntOrNull() ?: HttpStatusCode.Unauthorized.value
```

**Review by osipxd — CHANGES_REQUESTED** [review]

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt
Comment (wkornewald) [reply ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt:40]: But then you couldn't build on top of existing values:

```kotlin
val original = isUnauthorized
isUnauthorized = { original(it) || ... }
```

Comment (osipxd) [reply ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt:40]: We could make only `set()` private then, giving the possibility to get the configured `isAuthorized`:
```kotlin
public var isUnauthorized: suspend (HttpResponse) -> Boolean = { it.status == HttpStatusCode.Unauthorized }
    private set()
```
The point is to specify this field without an assignment operator.

File: buildSrc/src/main/kotlin/test/server/tests/Auth.kt
Comment (osipxd) [reply buildSrc/src/main/kotlin/test/server/tests/Auth.kt:132]: I thought that the value here should be equal to the value used in the default `isUnauthorized` implementation.
It is a very minor comment, you can ignore it :)

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt
Comment (wkornewald) [reply ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt:40]: Why is a getter + a differently named setter function better than an assignment operator unifying those two?

Comment (wkornewald) [reply ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt]: Please take a look at the latest commit. I've changed it. Though, this solution seems less direct and understandable, especially when there's a public getter (which probably is missing in many of the other APIs?). The `reAuthorizeOnStatuses` extra function can also still be provided with a var, too, so I don't see that argument.

**Review by osipxd — APPROVED** [review]
Thank you for keeping this branch up-to-date!

Could you [allow editing by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)? I want to mark `isUnauthorizedResponse` as `@InternalAPI`.

**Comment by wkornewald** [comment]
Oh right I forgot the InternalAPI markers. You should have access now.

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt
Comment (wkornewald) [inline ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt:45]: Why is this an internal API? My intention was to allow access to the current value, so you can extend an existing rule.

Comment (osipxd) [reply ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/Auth.kt:45]: We want to preserve the right to introduce breaking changes to this field. Keeping it public for those who’re ready to further breaking changes.

**Pull request merged**
